src/DataProvider/OperationDataProviderTrait.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\DataProvider;
+
+use ApiPlatform\Core\Exception\InvalidIdentifierException;
+use ApiPlatform\Core\Identifier\Normalizer\ChainIdentifierDenormalizer;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+trait OperationDataProviderTrait
+{
+    /**
+     * @var CollectionDataProviderInterface
+     */
+    private $collectionDataProvider;
+    /**
+     * @var ItemDataProviderInterface
+     */
+    private $itemDataProvider;
+    /**
+     * @var SubresourceDataProviderInterface
+     */
+    private $subresourceDataProvider;
+    /**
+     * @var ChainIdentifierDenormalizer
+     */
+    private $identifierDenormalizer;
+
+    /**
+     * Retrieves data for a collection operation.
+     *
+     * @return array|\Traversable|null
+     */
+    private function getCollectionData(Request $request, array $attributes, array $context)
+    {
+        if ($request->isMethod('POST')) {
+            return null;
+        }
+
+        return $this->collectionDataProvider->getCollection($attributes['resource_class'], $attributes['collection_operation_name'], $context);
+    }
+
+    /**
+     * Gets data for an item operation.
+     *
+     * @throws NotFoundHttpException
+     *
+     * @return object|null
+     */
+    private function getItemData(Request $request, array $attributes, array $context)
+    {
+        $id = $request->attributes->get('id');
+        $context = [];
+
+        try {
+            if ($this->identifierDenormalizer) {
+                $id = $this->identifierDenormalizer->denormalize((string) $id, $attributes['resource_class']);
+                $context = [ChainIdentifierDenormalizer::HAS_IDENTIFIER_DENORMALIZER => true];
+            }
+
+            $data = $this->itemDataProvider->getItem($attributes['resource_class'], $id, $attributes['item_operation_name'], $context);
+        } catch (InvalidIdentifierException $e) {
+            $data = null;
+        }
+
+        if (null === $data) {
+            throw new NotFoundHttpException('Not Found');
```

Comment:
We shouldn't throw a HTTP error here, but a domain exception.

Reply:
totally should be in ReadListener only